eras/conway/impl/src/Cardano/Ledger/Conway/Governance.hs

```diff
@@ -1126,6 +1108,71 @@ extractDRepPulsingState :: DRepPulsingState era -> RatifyState era
 extractDRepPulsingState x@(DRPulsing _) = snd (finishDRepPulser x)
 extractDRepPulsingState (DRComplete _ x) = x
 
+setCompleteDRepPulsingState ::
+  GovState era ~ ConwayGovState era =>
+  PulsingSnapshot era ->
+  RatifyState era ->
+  EpochState era ->
+  EpochState era
+setCompleteDRepPulsingState snapshot ratifyState epochState =
+  epochState
+    & epochStateGovStateL . cgDRepPulsingStateL
+      .~ DRComplete snapshot ratifyState
+
+-- | Refresh the pulser in the EpochState using all the new data that is needed to compute
+-- the RatifyState when pulsing completes.
+setFreshDRepPulsingState ::
+  ( GovState era ~ ConwayGovState era
+  , Monad m
+  , RunConwayRatify era
+  ) =>
+  EpochNo ->
+  PoolDistr (EraCrypto era) ->
+  EpochState era ->
+  ReaderT Globals m (EpochState era)
+setFreshDRepPulsingState epochNo stakePoolDistr epochState = do
+  -- When we are finished with the pulser that was started at the last epoch boundary, we
+  -- need to initialize a fresh DRep pulser. We do so by computing the pulse size and
+  -- gathering the data, which we will snapshot inside the pulser. We expect approximately
+  -- 10*k-many blocks to be produced each epoch, where `k` value is the stability
+  -- window. We must ensure for secure operation of the Hard Fork Combinator that we have
+  -- the new EnactState available 2 stability windows before the end of the epoch, while
+  -- spreading out stake distribution computation throughout the first 8 stability
+  -- windows. Therefore, we divide the number of stake credentials by 8*k
+  globals <- ask
+  let ledgerState = epochState ^. esLStateL
+      utxoState = lsUTxOState ledgerState
+      stakeDistr = credMap $ utxosStakeDistr utxoState
+      certState = lsCertState ledgerState
+      dState = certDState certState
+      vState = certVState certState
+      govState = epochState ^. epochStateGovStateL
+      stakeSize = Map.size stakeDistr
+      -- Maximum number of blocks we are allowed to roll back
```

Comment:
I guess it makes some sense to pass the entire epochstate to this function instead of the individual  pieces needed because we want to be sure that we have made all the changes to the epoch state in the EPOCH rule before we run this.
But we lose all knowledge (in the Rule where this is called) of what the snap shots are.
Just a suggestion, bu maybe define a type Snapshots, and a function getSnapshots :: EpochState -> Snapshots, then use a sequence like
let snap = getSpnaps epochstate
setFreshDRepPulsingState snap




























......

Reply:
> But we lose all knowledge (in the Rule where this is called) of what the snap shots are.

Do we need that in the EPOCH rule? This is exactly what I am trying to avoid in this PR. That rule is already huge and complex, we don't need to clutter it with any more logic. This is a specialized function that captures all of this information that is being snapshotted, which a reader can always inspect. I really don't get the argument why we need a whole separate type (that would only make it more confusing with another unrelated `Snapshot` type in Shelley). Especially since it can never be used for anything except with `runConwayRatify`.

I want to avoid scattering this information around different modules. This was one of the confusing parts of previous implementation: in one module pulser is constructed and in another all information needed for the pulser is gathered.